core/src/impl/Kokkos_SharedAlloc.hpp

```diff
@@ -24,6 +24,18 @@
 #include <cstdint>
 #include <string>
 
+// Use likely, unlikely attributes from C++20 to improve performance of
+// reference counting when copy constructors are involved.
+#if defined(KOKKOS_ENABLE_CXX20)
+#if defined(KOKKOS_IMPL_REF_COUNT_BRANCH_LIKELY)
+#define KOKKOS_BRANCH_PROB [[likely]]
+#elif defined(KOKKOS_IMPL_REF_COUNT_BRANCH_UNLIKELY)
+#define KOKKOS_BRANCH_PROB [[unlikely]]
+#endif
+#else
+#define KOKKOS_BRANCH_PROB
+#endif
```

Comment:
You have two if-conditions but only have an else branch for one of them. You want
```suggestion
#if defined(KOKKOS_ENABLE_CXX20)
#if defined(KOKKOS_IMPL_REF_COUNT_BRANCH_LIKELY)
#define KOKKOS_BRANCH_PROB [[likely]]
#elif defined(KOKKOS_IMPL_REF_COUNT_BRANCH_UNLIKELY)
#define KOKKOS_BRANCH_PROB [[unlikely]]
#endif
#endif
#ifndef KOKKOS_BRANCH_PROP
#define KOKKOS_BRANCH_PROP
#endif
```
or
```suggestion
#if defined(KOKKOS_ENABLE_CXX20)
#if defined(KOKKOS_IMPL_REF_COUNT_BRANCH_LIKELY)
#define KOKKOS_BRANCH_PROB [[likely]]
#elif defined(KOKKOS_IMPL_REF_COUNT_BRANCH_UNLIKELY)
#define KOKKOS_BRANCH_PROB [[unlikely]]
#else #define KOKKOS_BRANCH_PROP
#endif
#else #define KOKKOS_BRANCH_PROP
#endif
```

Reply:
Ultimately, we are only considering `[[unlikely]]` anyway, right?

Reply:
Oh right. Fixed it. Thanks.

Reply:
> Ultimately, we are only considering `[[unlikely]]` anyway, right?

Yes. That has been the best choice in our performance tests.